bin/shjs

```diff
@@ -1,12 +1,22 @@
 #!/usr/bin/env node
-require('../global');
 
-if (process.argv.length < 3) {
-  console.log('ShellJS: missing argument (script name)');
+if (require.main !== module) {
+  throw new Error('Executable-only module should not be required');
+}
+
+function exitWithErrorMessage(msg) {
+  console.log(msg);
   console.log();
   process.exit(1);
 }
 
+if (process.argv.length < 3) {
+  exitWithErrorMessage('ShellJS: missing argument (script name)');
+}
+
+// we only need ShellJS methods past this point, so require here
```

Comment:
Our usual style is to import at the top, not to import immediately before use. If there's a real requirement that we import down here, please document that requirement.

I suspect the requirement is: "we must import this after the `require.main` check, otherwise we pollute the global namespace as a side effect of the `disallow require-ing` test case."

If that's the real requirement, then you should move this immediately below the `require.main` check (and indicate this requirement). Otherwise, please explain why this is the appropriate spot, because I can't figure that out.

Reply:
Correct, the `require` at least needs to be after the `require.main` check, because the error it throws is catchable, and we don't want any side effects, as you said. I also moved it after the `process.argv` check, because we still don't need any globals at that point, but the fail condition is exiting. There is less of a benefit to `require`-ing after this check, so I will move it up.